File: Designer.Templates/Backtest/sma_strategy.py
import clr

clr.AddReference("System.Drawing")
clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from System import Math
from System.Drawing import Color
from StockSharp.Messages import UnitTypes
from StockSharp.Messages import Unit
from StockSharp.Messages import DataType
from StockSharp.Messages import ICandleMessage
from StockSharp.Messages import CandleStates
from StockSharp.Messages import Sides
from StockSharp.Algo.Indicators import SMA
from StockSharp.Algo.Strategies import Strategy
from StockSharp.Algo.Strategies import ISubscriptionHandler

class sma_strategy(Strategy):
    """
    Sample strategy demonstrating the work with SMA indicators.
    
    See more examples: https://github.com/StockSharp/AlgoTrading
    """
    def __init__(self):
        super(sma_strategy, self).__init__()
        self._isShortLessThenLong = None

        # Initialize strategy parameters
        self._candleTypeParam = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(1))) \
            .SetDisplay("Candle type", "Candle type for strategy calculation.", "General")

        self._long = self.Param("Long", 80)
        self._short = self.Param("Short", 30)

        self._takeValue = self.Param("TakeValue", Unit(0, UnitTypes.Absolute))
        self._stopValue = self.Param("StopValue", Unit(2, UnitTypes.Percent))

    @property
    def CandleType(self):
        return self._candleTypeParam.Value

    @CandleType.setter
    def CandleType(self, value):
        self._candleTypeParam.Value = value

    @property
    def Long(self):
        return self._long.Value

    @Long.setter
    def Long(self, value):
        self._long.Value = value

    @property
    def Short(self):
        return self._short.Value

    @Short.setter
    def Short(self, value):
        self._short.Value = value

    @property
    def TakeValue(self):
        return self._takeValue.Value

    @TakeValue.setter
    def TakeValue(self, value):
        self._takeValue.Value = value

    @property
    def StopValue(self):
        return self._stopValue.Value

    @StopValue.setter
    def StopValue(self, value):
        self._stopValue.Value = value

    def OnReseted(self):
        """
        Resets internal state when strategy is reset.
        """
        super(sma_strategy, self).OnReseted()
        self._isShortLessThenLong = None

    def OnStarted(self, time):
        """
        Called when the strategy starts. Sets up indicators, subscriptions, charting, and protection.
        
        :param time: The time when the strategy started.
        """
        super(sma_strategy, self).OnStarted(time)

        # Create indicators
        longSma = SMA()
        longSma.Length = self.Long
        shortSma = SMA()
        shortSma.Length = self.Short

        # Bind candles set and indicators
        subscription = self.SubscribeCandles(self.CandleType)
        # Bind indicators to the candles and start processing
        subscription.Bind(longSma, shortSma, self.OnProcess).Start()

        # Configure chart if GUI is available
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, shortSma, Color.Coral)
            self.DrawIndicator(area, longSma)
            self.DrawOwnTrades(area)

        # Configure position protection
        self.StartProtection(self.TakeValue, self.StopValue)

    def OnProcess(self, candle, longValue, shortValue):
        """
        Processes each finished candle, logs information, and executes trading logic on SMA crossing.
        
        :param candle: The processed candle message.
        :param longValue: The current value of the long SMA.
        :param shortValue: The current value of the short SMA.
        """
        self.LogInfo("New candle {0}: {6} {1};{2};{3};{4}; volume {5}", candle.OpenTime, candle.OpenPrice, candle.HighPrice, candle.LowPrice, candle.ClosePrice, candle.TotalVolume, candle.SecurityId)

        # If candle is not finished, do nothing
        if candle.State != CandleStates.Finished:
            return

        # Determine if short SMA is less than long SMA
        isShortLessThenLong = shortValue < longValue

        if self._isShortLessThenLong is None:
            self._isShortLessThenLong = isShortLessThenLong
        elif self._isShortLessThenLong != isShortLessThenLong:
            # Crossing happened
            direction = Sides.Sell if isShortLessThenLong else Sides.Buy

            # Calculate volume for opening position or reverting
            volume = self.Volume if self.Position == 0 else Math.Min(Math.Abs(self.Position), self.Volume) * 2

            # Get price step (default to 1 if not set)
            priceStep = self.GetSecurity().PriceStep or 1

            # Calculate order price with offset
            price = candle.ClosePrice + (priceStep if direction == Sides.Buy else -priceStep)

            if direction == Sides.Buy:
                self.BuyLimit(price, volume)
            else:
                self.SellLimit(price, volume)

            # Update state
            self._isShortLessThenLong = isShortLessThenLong

    def CreateClone(self):
        """
        !! REQUIRED!! Creates a new instance of the strategy.
        """
        return sma_strategy()

File: SamplesAvalonia/07_Testing/01_History/SmaStrategy.cs
namespace StockSharp.Samples.Testing.History.Avalonia;

using System;

using Ecng.Common;
using Ecng.ComponentModel;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Localization;
using StockSharp.Messages;

internal sealed class SmaStrategy : Strategy
{
	private bool? _isShortLessThanLong;
	private readonly StrategyParam<TimeSpan?> _candleTimeFrame;
	private readonly StrategyParam<int> _longSma;
	private readonly StrategyParam<int> _shortSma;
	private readonly StrategyParam<DataType> _candleType;
	private readonly StrategyParam<DataType> _buildFrom;
	private readonly StrategyParam<Level1Fields?> _buildField;
	private readonly StrategyParam<Unit> _takeValue;
	private readonly StrategyParam<Unit> _stopValue;

	public SmaStrategy()
	{
		_longSma = Param(nameof(LongSma), 80)
			.SetCanOptimize(true)
			.SetOptimize(50, 100, 5);
		_shortSma = Param(nameof(ShortSma), 30)
			.SetCanOptimize(true)
			.SetOptimize(20, 40, 1);
		_takeValue = Param(nameof(TakeValue), new Unit(0, UnitTypes.Absolute));
		_stopValue = Param(nameof(StopValue), new Unit(2, UnitTypes.Percent));
		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(1).TimeFrame()).SetRequired();
		_candleTimeFrame = Param<TimeSpan?>(nameof(CandleTimeFrame))
			.SetCanOptimize(true)
			.SetOptimize(TimeSpan.FromMinutes(5), TimeSpan.FromMinutes(15), TimeSpan.FromMinutes(5));
		_buildFrom = Param<DataType>(nameof(BuildFrom));
		_buildField = Param<Level1Fields?>(nameof(BuildField));
	}

	public TimeSpan? CandleTimeFrame
	{
		get => _candleTimeFrame.Value;
		set => _candleTimeFrame.Value = value;
	}

	public int LongSma
	{
		get => _longSma.Value;
		set => _longSma.Value = value;
	}

	public int ShortSma
	{
		get => _shortSma.Value;
		set => _shortSma.Value = value;
	}

	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	public DataType BuildFrom
	{
		get => _buildFrom.Value;
		set => _buildFrom.Value = value;
	}

	public Level1Fields? BuildField
	{
		get => _buildField.Value;
		set => _buildField.Value = value;
	}

	public Unit TakeValue
	{
		get => _takeValue.Value;
		set => _takeValue.Value = value;
	}

	public Unit StopValue
	{
		get => _stopValue.Value;
		set => _stopValue.Value = value;
	}

	protected override void OnReseted()
	{
		base.OnReseted();
		_isShortLessThanLong = null;
	}

	protected override void OnStarted2(DateTime time)
	{
		base.OnStarted2(time);

		var candleDataType = CandleTimeFrame is null
			? CandleType
			: DataType.Create(CandleType.MessageType, CandleTimeFrame);
		var subscription = new Subscription(candleDataType, Security)
		{
			MarketData =
			{
				IsFinishedOnly = true,
				BuildFrom = BuildFrom,
				BuildMode = BuildFrom is null ? MarketDataBuildModes.LoadAndBuild : MarketDataBuildModes.Build,
				BuildField = BuildField,
			},
		};

		var longSma = new SMA { Length = LongSma };
		var shortSma = new SMA { Length = ShortSma };
		SubscribeCandles(subscription)
			.Bind(longSma, shortSma, OnProcess)
			.Start();

		var area = CreateChartArea();
		if (area is not null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, shortSma, System.Drawing.Color.Coral);
			DrawIndicator(area, longSma);
			DrawOwnTrades(area);
		}

		StartProtection(TakeValue, StopValue);
	}

	private void OnProcess(ICandleMessage candle, decimal longValue, decimal shortValue)
	{
		LogInfo(
			LocalizedStrings.SmaNewCandleLog,
			candle.OpenTime,
			candle.OpenPrice,
			candle.HighPrice,
			candle.LowPrice,
			candle.ClosePrice,
			candle.TotalVolume,
			candle.SecurityId);

		if (candle.State != CandleStates.Finished)
			return;

		var isShortLessThanLong = shortValue < longValue;
		if (_isShortLessThanLong is null)
		{
			_isShortLessThanLong = isShortLessThanLong;
			return;
		}

		if (_isShortLessThanLong == isShortLessThanLong)
			return;

		var direction = isShortLessThanLong ? Sides.Sell : Sides.Buy;
		var volume = Position == 0 ? Volume : Position.Abs().Min(Volume) * 2;
		if (direction == Sides.Buy)
			BuyLimit(candle.ClosePrice, volume);
		else
			SellLimit(candle.ClosePrice, volume);

		_isShortLessThanLong = isShortLessThanLong;
	}
}
